### ISSUE TITLE
uniq-by-line: the surviving issue is chosen by the linter's alphabetical name

### ISSUE THREAD

**Issue by Semih702** [golangci/golangci-lint#6718]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official binary

### Description of the problem

When two linters report on the same line, `uniq-by-line` (default `true`) keeps one and drops the other. Which one survives is decided by the linter's name, alphabetically.

In the example below both linters report on line 6:

- `noctx` — use `QueryContext` instead of `Query` (a style suggestion)
- `rowserrcheck` — `rows.Err()` is never checked (a real bug: if the query fails mid-stream the function returns a truncated slice together with a nil error)

Only `noctx` is reported. The `rowserrcheck` finding is dropped, because `noctx` sorts before `rowserrcheck`. The verbose output shows both the ordering and the drop:

```
[lintersdb] Active 2 linters: [noctx rowserrcheck]
[runner] Processors filtering stat (in/out): ... uniq_by_line: 2/1 ...
```

With `uniq-by-line: false` both are reported.

The result flips with the names: when `bodyclose` and `noctx` land on the same line, `bodyclose` wins and the bug is reported. So whether a bug is surfaced depends on how the linter's name happens to sort.

I don't think this is intentional. Linters are sorted by name in `lintersdb`, and the comment on that sort says it is there to "make order of execution of linters (go/analysis metalinter and unused) stable". `UniqByLine` keeps the first issue per line, so it inherits that order as a priority it was never meant to be. `SortResults` runs later in the pipeline, so at dedup time there is no severity or category information available.

I'm not arguing against the dedup itself, only that the tie-break should be meaningful. Some directions:

1. order issues by something explicit before `uniq_by_line`
2. only dedup within a single linter, not across different ones
3. make the preference configurable

Happy to open a PR for whichever of these you'd accept.

Separately, and much smaller: `uniq_by_line` is the only one of the limiting processors that drops issues without saying so. `max_same_issues` and `max_from_linter` both log `"N/M issues ... were hidden, use --flag"` in `Finish()`. I'll send a small PR to make `uniq_by_line` do the same, which at least makes the current behaviour discoverable with `-v`.

For context on why the default matters: of 13 large Go projects I checked (Prometheus, Grafana, Docker CLI, moby, Traefik, Gitea, containerd, Cilium, Vault, MinIO, Argo CD, Rancher, Jaeger), none set `uniq-by-line`, so all of them run with this behaviour.

### Version of golangci-lint

```
golangci-lint has version 2.12.2 built with go1.26.2 from c0d3ddc9 on 2026-05-06T11:07:58Z
```

### Configuration

```yml
version: "2"
linters:
  default: none
  enable:
    - noctx
    - rowserrcheck
```

### Go environment

<details>

```
go version go1.25.6 windows/amd64
set AR=ar
set CC=gcc
set CGO_CFLAGS=-O2 -g
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-O2 -g
set CGO_ENABLED=1
set CGO_FFLAGS=-O2 -g
set CGO_LDFLAGS=-O2 -g
set CXX=g++
set GCCGO=gccgo
set GO111MODULE=
set GOAMD64=v1
set GOARCH=amd64
set GOAUTH=netrc
set GOBIN=
set GOCACHE=C:\Users\semih\AppData\Local\go-build
set GOCACHEPROG=
set GODEBUG=
set GOENV=C:\Users\semih\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFIPS140=off
set GOFLAGS=
set GOGCCFLAGS=-m64 -mthreads -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=C:\Users\semih\AppData\Local\Temp\go-build1135275179=/tmp/go-build -gno-record-gcc-switches
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMOD=C:\Users\semih\AppData\Local\Temp\claude\D--go-patterns\4cfe6c6f-adaa-4bfd-af6d-aabaf17f4848\scratchpad\minrepro\go.mod
set GOMODCACHE=C:\Users\semih\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\Users\semih\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=D:\Programming\Go
set GOSUMDB=sum.golang.org
set GOTELEMETRY=local
set GOTELEMETRYDIR=C:\Users\semih\AppData\Roaming\go\telemetry
set GOTMPDIR=
set GOTOOLCHAIN=auto
set GOTOOLDIR=D:\Programming\Go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.25.6
set GOWORK=
set PKG_CONFIG=pkg-config
```

</details>

### Verbose output of running

<details>

```
level=info msg="golangci-lint has version 2.12.2 built with go1.26.2 from c0d3ddc9 on 2026-05-06T11:07:58Z"
level=info msg="[config_reader] Config search paths: [./ C:\\Users\\semih\\AppData\\Local\\Temp\\claude\\D--go-patterns\\4cfe6c6f-adaa-4bfd-af6d-aabaf17f4848\\scratchpad\\minrepro C:\\Users\\semih\\AppData\\Local\\Temp\\claude\\D--go-patterns\\4cfe6c6f-adaa-4bfd-af6d-aabaf17f4848\\scratchpad C:\\Users\\semih\\AppData\\Local\\Temp\\claude\\D--go-patterns\\4cfe6c6f-adaa-4bfd-af6d-aabaf17f4848 C:\\Users\\semih\\AppData\\Local\\Temp\\claude\\D--go-patterns C:\\Users\\semih\\AppData\\Local\\Temp\\claude C:\\Users\\semih\\AppData\\Local\\Temp C:\\Users\\semih\\AppData\\Local C:\\Users\\semih\\AppData C:\\Users\\semih C:\\Users C:\\]"
level=info msg="[config_reader] Used config file .golangci.yml"
level=info msg="[config_reader] Module name \"repro\""
level=info msg="[goenv] Read go env for 56.4938ms: map[string]string{\"GOCACHE\":\"C:\\\\Users\\\\semih\\\\AppData\\\\Local\\\\go-build\", \"GOROOT\":\"D:\\\\Programming\\\\Go\"}"
level=info msg="[lintersdb] Active 2 linters: [noctx rowserrcheck]"
level=info msg="[loader] Go packages loading at mode 8767 (exports_file|files|imports|name|compiled_files|deps|types_sizes) took 386.5232ms"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 551.7µs"
level=info msg="[linters_context/goanalysis] analyzers took 378.7873ms with top 10 stages: ctrlflow: 355.6331ms, inspect: 23.1542ms, typecheck: 0s, buildssa: 0s, noctx: 0s, rowserrcheck: 0s"
level=info msg="[runner] Issues before processing: 2, after processing: 1"
level=info msg="[runner] Processors filtering stat (in/out): path_shortener: 1/1, path_absoluter: 2/2, cgo: 2/2, invalid_issue: 2/2, generated_file_filter: 2/2, uniq_by_line: 2/1, max_same_issues: 1/1, severity-rules: 1/1, exclusion_paths: 2/2, exclusion_rules: 2/2, fixer: 2/2, max_per_file_from_linter: 1/1, max_from_linter: 1/1, source_code: 1/1, path_prettifier: 1/1, sort_results: 1/1, filename_unadjuster: 2/2, path_relativity: 2/2, nolint_filter: 2/2, diff: 2/2"
level=info msg="[runner] processing took 642.9µs with stages: source_code: 642.9µs, path_relativity: 0s, nolint_filter: 0s, diff: 0s, uniq_by_line: 0s, generated_file_filter: 0s, invalid_issue: 0s, path_prettifier: 0s, path_absoluter: 0s, exclusion_paths: 0s, cgo: 0s, filename_unadjuster: 0s, fixer: 0s, path_shortener: 0s, severity-rules: 0s, exclusion_rules: 0s, max_per_file_from_linter: 0s, max_same_issues: 0s, max_from_linter: 0s, sort_results: 0s"
level=info msg="[runner] linters took 612.518ms with stages: goanalysis_metalinter: 611.8751ms"
repro.go:6:23: (*database/sql.DB).Query must not be called. use (*database/sql.DB).QueryContext (noctx)
	rows, err := db.Query("select name from users")
	                     ^
1 issues:
* noctx: 1
level=info msg="File cache stats: 1 entries of total size 380B"
level=info msg="Memory: 12 samples, avg is 79.7MB, max is 159.2MB"
level=info msg="Execution took 1.0560867s"
```

</details>

### A minimal reproducible example or link to a public repository

`repro.go`:

```go
package repro

import "database/sql"

func Names(db *sql.DB) ([]string, error) {
	rows, err := db.Query("select name from users")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []string
	for rows.Next() {
		var n string
		if err := rows.Scan(&n); err != nil {
			return nil, err
		}
		out = append(out, n)
	}
	return out, nil // rows.Err() is never checked
}
```

`golangci-lint run` reports only `noctx`. Adding `issues: {uniq-by-line: false}` reports both.

### Validation

- [x] Yes, I've included all the information above (example, version, config, etc.).

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.

**Comment by ldez** [comment]
This is not a bug, this is expected behavior.
